drivers/spi/spi_silabs_eusart.c

```diff
@@ -152,6 +217,330 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+}
+
+static void spi_silabs_build_dma_config(const struct device *dev,
+					struct dma_config *current_dma_config,
+					const struct dma_block_config *blk_cfg,
+					uint32_t block_count, bool is_tx)
+{
+	memset(current_dma_config, 0, sizeof(*current_dma_config));
+
+	if (is_tx) {
+		/* Memory to peripheral. */
+		current_dma_config->channel_direction = MEMORY_TO_PERIPHERAL;
+	} else {
+		/* Peripheral to memory. */
+		current_dma_config->channel_direction = PERIPHERAL_TO_MEMORY;
+	}
+
+	/* invoke callback at the end of transfer list. */
+	current_dma_config->complete_callback_en = 0x00U;
+
+	current_dma_config->source_data_size = 0x01U; /* 1 byte. */
+	current_dma_config->dest_data_size = 0x01U;   /* 1 byte. */
+
+	current_dma_config->source_burst_length = 0x01U;
+	current_dma_config->dest_burst_length = 0x01U;
+
+	current_dma_config->block_count = block_count;
+
+	current_dma_config->head_block = (struct dma_block_config *)blk_cfg;
+
+	current_dma_config->user_data = (void *)dev;
+
+	if (!is_tx) {
+		/* Only enable the end transaction callback. Only SPI controller mode is supported.
+		 */
+		current_dma_config->dma_callback = &spi_silabs_dma_rx_callback;
+	}
+}
+
+static int spi_silabs_prepare_dma_transfer(const struct device *dev,
+					   size_t total_padded_transaction_size, bool is_tx)
+{
+	struct dma_block_config *current_blk_cfg = NULL;
+	const struct spi_buf *current_buffer = NULL;
+	struct spi_silabs_eusart_data *instance_data = dev->data;
+	const struct spi_silabs_eusart_config *instance_conf = dev->config;
+
+	struct dma_config dma_cfg;
+
+	size_t current_total_transfer_size = 0U;
+	size_t padding_size = 0U;
+	size_t buffer_count = 0U;
+	size_t dma_block_config_idx = 0U;
+	int ret = 0U;
+
+	if (is_tx) {
+		current_buffer = &instance_data->ctx.current_tx[0];
+		buffer_count = instance_data->ctx.tx_count;
+		current_blk_cfg = &(instance_data->dma_channel_tx.dma_descriptors[0]);
+	} else {
+		current_buffer = &instance_data->ctx.current_rx[0];
+		buffer_count = instance_data->ctx.rx_count;
+		current_blk_cfg = &(instance_data->dma_channel_rx.dma_descriptors[0]);
+	}
+
+	/* For each buffer in the instance context buffer set. */
+	for (size_t i = 0U; i < buffer_count; i++) {
+		size_t buffer_remaining_count = current_buffer->len;
+
+		current_total_transfer_size += current_buffer->len;
+
+		if (buffer_remaining_count == 0U) {
+			/* Skip zero length buffers */
+			continue;
+		}
+
+		/* Create the required number of dma transfer block for this buffer. */
+		while (dma_block_config_idx < SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+			uint32_t index_in_buffer = current_buffer->len - buffer_remaining_count;
+
+			buffer_remaining_count = spi_silabs_build_dma_transfer_block(
+				instance_conf, &(current_blk_cfg[dma_block_config_idx]),
+				&(((uint8_t *)current_buffer->buf)[index_in_buffer]),
+				buffer_remaining_count, is_tx);
+
+			dma_block_config_idx++;
+
+			if (dma_block_config_idx > SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
```

Comment:
Can this condition occur? The while loop runs while `<`, and the variable is only incremented once. Seems like this check should be for `==`, and happen at the beginning of the loop, since `==` isn't an error condition when checked _after_ incrementing.